plugins/catalog-backend/src/ingestion/processors/GitlabReaderProcessor.ts

```diff
@@ -15,11 +15,128 @@
  */
 
 import { LocationSpec } from '@backstage/catalog-model';
-import fetch from 'node-fetch';
+import { Config } from '@backstage/config';
+import parseGitUri from 'git-url-parse';
+import fetch, { HeadersInit, RequestInit } from 'node-fetch';
+import { Logger } from 'winston';
 import * as result from './results';
 import { LocationProcessor, LocationProcessorEmit } from './types';
 
+/**
+ * The configuration parameters for a single GitLab API provider.
+ */
+export type GitLabProviderConfig = {
+  /**
+   * The prefix of the target that this matches on, e.g. "https://gitlab.com",
+   * with no trailing slash.
+   */
+  target: string;
+
+  /**
+   * The base URL of the API of this provider, e.g. "https://api.gitlab.com",
+   * with no trailing slash.
+   *
+   * The API will always be preferred if both its base URL and a token are
+   * present.
+   */
+  apiBaseUrl?: string;
+
+  /**
+   * The authorization token to use for requests to this provider.
+   *
+   * If no token is specified, anonymous access is used.
+   */
+  token?: string;
+};
+
+export function readConfig(
+  config: Config,
+  logger: Logger,
+): GitLabProviderConfig[] {
+  const providers: GitLabProviderConfig[] = [];
+
+  const providerConfigs =
+    config.getOptionalConfigArray('catalog.processors.gitlab.providers') ?? [];
+
+  // First read all the explicit providers
+  for (const providerConfig of providerConfigs) {
+    const target = providerConfig.getString('target').replace(/\/+$/, '');
+    let apiBaseUrl = providerConfig.getOptionalString('apiBaseUrl');
+    const token = providerConfig.getOptionalString('token');
+
+    if (apiBaseUrl) {
+      apiBaseUrl = apiBaseUrl.replace(/\/+$/, '');
+    } else if (target === 'https://gitlab.com') {
+      apiBaseUrl = 'https://gitlab.com/api/v4';
+    }
+
+    if (!apiBaseUrl) {
+      throw new Error(
+        `Provider at ${target} must configure an explicit apiBaseUrl`,
+      );
+    }
+
+    providers.push({ target, apiBaseUrl, token });
+  }
+
+  return providers;
+}
+
+// Converts for example
+// from: https://gitlab.example.com/organization/group/project/blob/branchname/path/to/c.yaml
+// to:   https://gitlab.example.com/api/v4/projects/13083/project/files/c.yaml?ref=branchname
+export function getApiUrl(target: string, provider: GitLabProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+    if (
+      !owner ||
+      !name ||
+      !ref ||
+      (filepathtype !== 'blob' && filepathtype !== 'raw') ||
+      !filepath?.match(/\.ya?ml$/)
+    ) {
+      throw new Error('Wrong URL or invalid file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+    return new URL(
+      `${provider.apiBaseUrl}/projects/${`${owner}/${name}`.replace(
+        /\//g,
+        '%2F',
```

Comment:
You should be able to use `encodeURIComponent` here instead so that it will escape the slashes for you